### ISSUE TITLE
Investigate right-to-left (RTL) support in Defold

### ISSUE THREAD

**Issue by britzl** [defold/defold#4900]
**Is your feature request related to a problem? Please describe (REQUIRED):**
Defold doesn't support RTL text which makes it hard to use RTL languages such as Persian, Arabic and Hebrew.

**Describe the solution you'd like (REQUIRED):**
The label and text node should have options to toggle between LTR and RTL, both in editor and at run-time using API methods.

**Describe alternatives you've considered (REQUIRED):**
There are workarounds discussed here: https://forum.defold.com/t/supporting-non-english-languages-and-rtl-languages/16991

Basically manually reversing the string etc

**Comment by hapass** [comment]
@britzl 
Is it going to be a simplified implementation? Because reversing the string is not enough for full RTL support. Keep in mind that we use Arabic numeral system and numbers should not be reversed. There are also usually pitfalls when reversing text with unicode surrogate pairs etc.

**Comment by ntorabi85** [comment]
not much problem with single line simple texts but multiline texts and those text that are mixe of LTR and RTL languages make different problems

**Comment by britzl** [comment]
@hapass I don't know yet how this will work. We'd need to design the solution and look at best practices (how does other games engines solve this?). It should be something more than simply reversing strings if possible.

**Comment by spritefullake** [comment]
Hello, curious if any progress has been made in this front or if there are any workarounds? I would be happy to help add Defold engine support but I don't know the internals too well myself. 

I found [this](https://forum.defold.com/t/solution-for-arabic-text/66496) but I haven't tested it out myself. There are also [these suggestions](https://forum.defold.com/t/android-app-development/64542/2). Are there any working examples of integrating these libraries?

Thank you.

**Comment by britzl** [comment]
@spritefullake I'm afraid no work has been done on RTL support yet.

> There are also [these suggestions](https://forum.defold.com/t/android-app-development/64542/2). Are there any working examples of integrating these libraries?

I believe each of these repositories contain Defold example project that you can try.

**Comment by TheUnderName** [comment]
In Godot, in order for RTL to be fixed, we first imported a Persian font like (Vazir.ttf) and copied the desired text using Text Converter sites like (https://www.timecode.ir/font-converter/) We used to do it and paste inside the editor

**Comment by TheUnderName** [comment]
https://github.com/rastikerdar/vazirmatn

**Comment by AGulev** [comment]
it's not the solution but a workaround for now 
https://forum.defold.com/t/supporting-non-english-languages-and-rtl-languages/16991/14?u=agulev

**Comment by TheUnderName** [comment]
Those who have my problem, you can activate the All Chars button by using Persian or Arabic fonts and convert to Persian or Arabic by using the recommended site: https://www.timecode.ir/font-converter/ Suggested font: Vazir

**Comment by britzl** [comment]
Changed this to an investigation task to begin with. We need to figure out actual implementation tasks first.

**Comment by TheUnderName** [comment]
Godot uses https://github.com/harfbuzz/harfbuzz for RTL

**Comment by karimboubouh** [comment]
I need to write Arabic text in my game. Is there any solution or at least a workaround ?

**Comment by ask-a-lie** [comment]
@karimboubouh 
Hi, I used this extension to render Hindi text and it works fine, so you can use it for Arabic text, I think.
This extension can render text using native label and retrieve it as pixel data.
https://github.com/Lerg/extension-nativetext 

However, this extension can't be used as is, so it requires some modifications.
1. Need to change [line 44 in extension.cpp](https://github.com/Lerg/extension-nativetext/blob/bf19acb09907cfd3afffaf6925038818a3b40a91/nativetext/src/extension.cpp#L44) as follows:
```diff
- dmScript::LuaHBuffer lua_buffer = {buffer, true};
+ dmScript::LuaHBuffer lua_buffer(buffer, dmScript::OWNER_LUA);
```
2. (Optional)Convert data to set texture for gui as follows:
```
function set_texture(node)
	-- Set maximum allowed texture size.
	nativetext.init {
		max_texture_width = 2048,
		max_texture_height = 2048
	}
	-- Generate text image.
	local image_resource = nativetext.new {
		text = 'This is a test ど러연关\nыыыьщюш\nตตต ตัตัตั ตั้ตั้ตั้',
		width = 250,
		size = 30,
		spacing_add = 50,
		align = nativetext.align_center,
		color = vmath.vector4(0, 1, 1, 1),
		outline_size = 8,
		outline_color = vmath.vector4(0.9, 0.1, 0.1, 1),
		shadow_size = 8,
		shadow_x = 8, shadow_y = -8,
		shadow_color = vmath.vector4(0, 0, 0, 1),
		background_color = vmath.vector4(0.1, 0.2, 0.8, 1)
	}
	-- Convert pixels to string for gui.new_texture
	local bytes = buffer.get_stream(image_resource.buffer, hash("pixels"))
	local pixels = {}
	for i = 1, #bytes do
		table.insert(pixels, string.char(bytes[i]))
	end

	local concatenated_pixels = table.concat(pixels)
	if gui.new_texture("texture_id", image_resource.header.width, image_resource.header.height, "rgba", concatenated_pixels, true) then
		gui.set_texture(node, "texture_id")
	else
		print("error")
	end
end
```

I recently discovered this method, so I hope it works well for you too 😃

**Comment by AGulev** [comment]
Copied from https://github.com/defold/defold/issues/7475

>I’ve tried different fonts, letter spacing and line spacing, but the result is the same.
Left-Defold, right - google font site:
![46f825761c661008eb7545e97837012715c2bab7](https://user-images.githubusercontent.com/21050233/225602885-a5a5d710-34da-4a67-9dc5-f830d4029e17.png)

**Comment by jalhund** [comment]
Hello, @britzl is this planned to be added in the near future?

**Comment by britzl** [comment]
> Hello, @britzl is this planned to be added in the near future?

It is not really super high on our list no. What kind of impact will that have for your current project?

**Comment by jalhund** [comment]
> It is not really super high on our list no. What kind of impact will that have for your current project?

Nothing critical, but native Arabic and Hebrew speakers complain about this. It would be cool to have good support for this in the engine itself.

I'll add one of the workarounds from this thread, but if I understand correctly, they may not handle some characters or combinations of them correctly.

**Comment by britzl** [comment]
We have finished a design for a new text rendering system: https://github.com/defold/defold/issues/9692

Implementation in: https://github.com/defold/defold/issues/9872
